### PR TITLE
Feature/asg term and node age roll

### PR DESCRIPTION
**Pull request by sarmad-abualkaz**
Adding a mode 4 to the script which aims to update (roll) nodes based on each ec2 instance age (e.g. older than 6 days).

This will have the exact same functionality expectation as mode 1 except for evaluating outdated nodes.

Another change is to also remove the requirement to call `terminate_instances` from EC2 client and instead performing termination via Autoscaling client using `terminate_instance_in_auto_scaling_group` which ensures IAM roles can be more narrow to an ASG ARN and also eliminate the need to detach instances from ASGs.

### CODE REVIEW

File: lib/aws.py [hunk 1]
```diff
@@ -32,26 +33,26 @@ def get_launch_template(lt_name):
...
-
+    if not app_config['DRY_RUN']:
+        logger.info('Terminating ec2 instance in ASG {}...'.format(instance_id))
+        try:
+            response = client.terminate_instance_in_auto_scaling_group(
+                InstanceId=instance_id,
+                ShouldDecrementDesiredCapacity=True
+            )
+            if response['ResponseMetadata']['HTTPStatusCode'] == requests.codes.ok:
+                logger.info('Termination singal for instance is successfully sent.')
```
Comment (hawkesn) [inline lib/aws.py]:
```suggestion
                logger.info('Termination signal for instance is successfully sent.')
```

File: Dockerfile [hunk 2]
```diff
@@ -19,4 +19,4 @@ RUN pip3 install --no-cache-dir -r requirements.txt
 COPY lib ./lib
 COPY config.py eks_rolling_update.py ./
 
-ENTRYPOINT ["python", "eks_rolling_update.py"]
+CMD ["python", "eks_rolling_update.py"]
```
Comment (andrewhertog) [inline Dockerfile]: Why is this change needed?

Comment (sarmad-abualkaz) [reply Dockerfile]: I reverted that having talked with @kuuji about it 😄

File: lib/aws.py [hunk 3]
```diff
@@ -32,26 +33,26 @@ def get_launch_template(lt_name):
...
+        logger.info('Terminating ec2 instance in ASG {}...'.format(instance_id))
+        try:
+            response = client.terminate_instance_in_auto_scaling_group(
+                InstanceId=instance_id,
+                ShouldDecrementDesiredCapacity=True
+            )
+            if response['ResponseMetadata']['HTTPStatusCode'] == requests.codes.ok:
+                logger.info('Termination signal for instance is successfully sent.')
+            else:
+                logger.info('Termination singal for instance has failed. Response code was {}. Exiting.'.format(response['ResponseMetadata']['HTTPStatusCode']))
```
Comment (kuuji) [inline lib/aws.py]: Typo `singal`

File: lib/aws.py [hunk 4]
```diff
@@ -256,6 +257,29 @@ def instance_outdated_launchtemplate(instance_obj, asg_lt_name, asg_lt_version):
...
+
+    instance_launch_time = response['Reservations'][0]['Instances'][0]['LaunchTime']
+
+    instance_age = ((datetime.datetime.now(instance_launch_time.tzinfo) - instance_launch_time).days)
+    instance_age_remainder = ((datetime.datetime.now(instance_launch_time.tzinfo) - instance_launch_time).seconds)
+
+    if instance_age > days_fresh:
+        logger.info("Instance id {} launch age of '{}' day(s) is older than expected '{}' day(s)".format(instance_id, instance_age, days_fresh))
+        return True
+    elif (instance_age == days_fresh) and (instance_age_remainder > 0):
```
Comment (kuuji) [inline lib/aws.py:279]: That logic is not very clear to read. 
I understand that you check the number of days and if it's the same you check the seconds. Why don't you just check the seconds to begin with? And just show the offset based on the result.

Either way I feel the code could be made a bit clearer. Maybe with a comment or by renaming these variable to carry what you're doing.

File: lib/aws.py [hunk 5]
```diff
@@ -256,6 +257,29 @@ def instance_outdated_launchtemplate(instance_obj, asg_lt_name, asg_lt_version):
     return False
 
 
+def older_instance (instance_id, days_fresh):
```
Comment (kuuji) [inline lib/aws.py]: Since this function returns True or False I would call it `is_instance_outdated`

File: lib/aws.py [hunk 6]
```diff
@@ -326,6 +350,30 @@ def plan_asgs(asgs):
 
     return asg_outdated_instance_dict
 
+def plan_asgs_older_nodes(asgs):
```
Comment (kuuji) [inline lib/aws.py:356]: Just naming again. I like to call my function exactly what they do. This gets you a list of outdated nodes. I'd call it `get_outdated_asg_nodes` or  `get_outdated_asg_instances` since you seem to be using instances in other places in the code.

Edit : I see that you're following their convention for plan. Maybe just stay consistent with naming then. `plan_asgs_outdated_instances` instead of older_nodes

File: eks_rolling_update.py [hunk 7]
```diff
@@ -181,12 +185,16 @@ def update_asgs(asgs, cluster_name):
...
                 delete_node(node_name)
-                terminate_instance(outdated['InstanceId'])
+                terminate_instance_in_asg(outdated['InstanceId'])
                 if not instance_terminated(outdated['InstanceId']):
                     raise Exception('Instance is failing to terminate. Cancelling out.')
-                detach_instance(outdated['InstanceId'], asg_name)
-                if not instance_detached(outdated['InstanceId']):
-                    raise Exception('Instance is failing to detach from ASG. Cancelling out.')
+
+                #terminate_instance(outdated['InstanceId'])
```
Comment (kuuji) [inline eks_rolling_update.py]: You should get rid of that commented code.

**Review by kuuji — CHANGES_REQUESTED** [review]
Nothing too big, mostly naming and typo and a bit of clarity.

Good job!

File: lib/aws.py (hunk 3 above)
Comment (sarmad-abualkaz) [reply lib/aws.py]: lol oh well...my parents still wonder how I finished pre-school 😂

File: lib/aws.py (hunk 4 above)
Comment (sarmad-abualkaz) [reply lib/aws.py:279]: Honestly @kuuji either I don't know how to use this library or it's not the appropriate library for this `datetime`, but somehow a difference in time is returning `xx days xxxxx seconds`.

Initially I was testing using only `.seconds` of the difference and it was missing nodes days older than I put as my limit, then I realized that was returning the `xxxxx seconds` of the time diff. 

So that's why I have something like `instance_age` which is the diff in days and when that's equal the  `days_fresh` EC2 age limit I look for the `instance_age_remainder` in seconds.

Do you feel comments here might make it clearer?

File: lib/aws.py (hunk 5 above)
Comment (sarmad-abualkaz) [reply lib/aws.py]: I ended-up following the same convention they used like  `instance_outdated_launchconfiguration`, `instance_outdated_launchtemplate`. This one I called `instance_outdated_age`. 

Maybe that's better?

File: lib/aws.py (hunk 4 above)
Comment (sarmad-abualkaz) [reply lib/aws.py:279]: I put comments on `instance_age` and `instance_age_remainder`, let me know if that clarifies the purpose of this?

**Review by kuuji — APPROVED** [review]
